cqlmapper/models.py

```diff
@@ -624,6 +625,117 @@ def timeout(self, timeout):
+            it's value must be given.
+
+            .. code-block:: python
+
+                class SimpleModel(Model):
+                    key = columns.Text(primary_key=True)
+                    value = columns.Text()
+
+                class ComplexModel(Model):
+                    pk = columns.Text(primary_key=True)
+                    ck = colums.Integer(primary_key=True)
+                    value = columns.Text()
+
+                valid_simple = SimpleModel.load_many(conn, ["fizz", "buzz"])
+                valid_simple = SimpleModel.load_many(conn, [{"key": "fizz"}, {"key, "buzz"}])
+                try:
+                    invalid_simple = SimpleModel.load_many(conn, ["fizz", {"key, "buzz"}])
+                except TypeError:
+                    pass
+
+                valid_complex = ComplexModel.load_many(
+                    conn=conn,
+                    keys=[
+                        {"pk": "fizz", "ck": "buzz},
+                        {"pk", "foo", "ck": "bar"},
+                    ],
+                )
+                try:
+                    invalid_complex = SimpleModel.load_many(conn, [{"pk": "fizz"}])
+                except ValueError:
+                    pass
+
+        :type: List[Dict[str, Any]] or List[Any]
+        :param concurrency: Maximum number of queries to run concurrently.
+        :type: int
+        """
+        if not keys:
+            return []
+
+        if concurrency < 1:
+            raise ValueError("'concurrency' in 'load_many' must be >= 1.")
+
+        # cls._primary_keys is an OrderedDict so no need to sort the keys
+        pks = list(cls._primary_keys.keys())
+
+        # Validate that 'keys' is either all dicts or all simple values for
+        # Models that allow that
+        simple_keys = False
+        parameters = []
+        for i, key_values in enumerate(keys):
+            if not isinstance(key_values, dict) and len(pks) == 1 and i == 0:
+                simple_keys = True
+
+            if simple_keys:
+                if isinstance(key_values, dict):
+                    raise TypeError(
+                        "'keys' has a mix of dicts and simple values, must be "
+                        "one or the other for model %s." % cls.__name__
+                    )
+                parameters.append((key_values,))
+            else:
+                if not isinstance(key_values, dict):
+                    raise TypeError(
+                        "All values in 'keys' must be a dict for model %s." % cls.__name__
+                    )
+                else:
+                    try:
+                        parameters.append(tuple(keys[key] for key in pks))
+                    except KeyError:
+                        raise ValueError(
+                            "An element in 'keys' is missing a PRIMARY KEY "
+                            "value for model %s." % cls.__name__
+                        )
```

Comment:
Should we go through all of these steps to try and verify that `keys` is correct before hand or just leave it be and let the query fail if someone does something wrong?

Reply:
does this make the error message significantly nicer? is this something that certain user-data would mess up but tests might not see? (so it's something we'd only catch at runtime?) or would it be like a programmer error that's cut and dried? I worry a bit about the extra overhead of doing all this if it's gonna fail every time / it's already pretty clear what went wrong.

Reply:
> does this make the error message significantly nicer?

It should, (although I'm not 100% sure _how_ it would fail currently).  I think, you could fail with a `KeyError` when generating the `parameters`, which isn't _too_ bad and the stack trace would probably help.  If you mix simple and dicts though, I think then you'll start seeing errors from the cassandra drivers which are not always especially helpful in debugging.

>  is this something that certain user-data would mess up but tests might not see?

This should _only_ be programmer error.

Reply:
This is how it looks when it's simplified:

https://github.com/reddit/cqlmapper/pull/20/files#diff-41b3c2d0a5cdb09762e9be1b018f36b7R684